fix: `meltano run` no longer creates empty `venv` plugin directories for inherited plugins

edgarrmondragon opened the pull request:
Related:

- Closes #8401

edgarrmondragon committed:
Add failing test

edgarrmondragon commented on tests/meltano/core/test_utils.py:
```diff
@@ -222,3 +223,22 @@ def test_remove_suffix():
+    cat = species("felis", "catus")
+    assert cat.exists()
+
+    wolf = species("canis", "lupus", make_dirs=False)
+    assert not wolf.exists()
```
Showcase of buggy behavior of `makedirs` before this fix:

```
        wolf = species("canis", "lupus", make_dirs=False)
>       assert not wolf.exists()
E       AssertionError: assert not True
E        +  where True = <bound method Path.exists of PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw0/test_makedirs_decorator0/canis/lupus')>()
E        +    where <bound method Path.exists of PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw0/test_makedirs_decorator0/canis/lupus')> = PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw0/test_makedirs_decorator0/canis/lupus').exists
```

that should've passed.

edgarrmondragon committed:
Fix venv path being created for inherited plugins

edgarrmondragon committed:
Fix TypeAlias annotation

ReubenFrankel commented:
We had a whole discussion about this back in the GitLab days: https://gitlab.com/meltano/meltano/-/merge_requests/2099#note_566826570

The change was even made to [add `**kwargs`](https://gitlab.com/meltano/meltano/-/merge_requests/2099/diffs?commit_id=9a90c502a426d2d1eff231994abe4e0ae61f141c), then [narrowed down to a `make_dirs` kwarg](https://gitlab.com/meltano/meltano/-/merge_requests/2099/diffs?commit_id=9a90c502a426d2d1eff231994abe4e0ae61f141c), then [this merge commit rolled back the change](https://gitlab.com/meltano/meltano/-/merge_requests/2099/diffs?commit_id=9a90c502a426d2d1eff231994abe4e0ae61f141c) and it went unnoticed. :cry: (I'm experimenting with a feature and am running into a problem with `makedirs` as well)

edgarrmondragon committed:
Use explicit `make_dirs` kwarg

edgarrmondragon commented:
Thanks for commenting @ReubenFrankel!

> We had a [whole discussion about this](https://gitlab.com/meltano/meltano/-/merge_requests/2099#note_566826570) back in the GitLab days - a change was even made to [add a `make_dirs` kwarg](https://gitlab.com/meltano/meltano/-/commit/c1aca6cce3e9f7e4d67dda555ecd1c23097e8a08#8e8067d772989bfee158f1ccfa6e2a8fe41056ec) to all `@makedirs` methods ([requested by AJ](https://gitlab.com/meltano/meltano/-/merge_requests/2099/diffs?commit_id=2d5af5e8ebd9faf655d056e1b0d2c09a09d4d526#note_569189837) after `**kwargs` was implemented initially)

Wow, I wasn't aware of that! And it makes it easier to decide a explicit kwarg over generic `kwargs` 😅.

> and then [78b009b#diff-c087424df56cc7c7f9cb7450ad5f6d97ad8855940e5b27b548c879948616140c](https://github.com/meltano/meltano/commit/78b009b4c195678d2178ff4e0b69c711fae4ffb9#diff-c087424df56cc7c7f9cb7450ad5f6d97ad8855940e5b27b548c879948616140c) was merged which reverted the changes without anyone noticing (probably no tests). 😢 I'm not sure if this is the case for anything else, but may be worth checking.

Yeah, that was a large merge and we missed it somehow 😞. Anyway this PR includes a test that should prevent some regressions.

edgarrmondragon committed:
Fix `plugin_lock_path` call

edgarrmondragon committed:
Fix tests

dependabot[bot] committed:
ci: bump the actions group with 3 updates (#8471)

Bumps the actions group with 3 updates: [actions/setup-python](https://github.com/actions/setup-python), [actions/dependency-review-action](https://github.com/actions/dependency-review-action) and [codecov/codecov-action](https://github.com/codecov/codecov-action).


Updates `actions/setup-python` from 5.0.0 to 5.1.0
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v5.0.0...v5.1.0)

Updates `actions/dependency-review-action` from 4.2.4 to 4.2.5
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/v4.2.4...v4.2.5)

Updates `codecov/codecov-action` from 4.1.0 to 4.1.1
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v4.1.0...v4.1.1)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: actions
- dependency-name: actions/dependency-review-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: actions
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: actions
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps-dev): bump the development-dependencies group with 5 updates (#8472)

Bumps the development-dependencies group with 5 updates:

| Package | From | To |
| --- | --- | --- |
| [boto3-stubs](https://github.com/youtype/mypy_boto3_builder) | `1.34.69` | `1.34.74` |
| [hypothesis](https://github.com/HypothesisWorks/hypothesis) | `6.99.13` | `6.100.0` |
| [moto](https://github.com/getmoto/moto) | `5.0.3` | `5.0.4` |
| [requests-mock](https://github.com/jamielennox/requests-mock) | `1.11.0` | `1.12.1` |
| [types-jsonschema](https://github.com/python/typeshed) | `4.21.0.20240311` | `4.21.0.20240331` |


Updates `boto3-stubs` from 1.34.69 to 1.34.74
- [Release notes](https://github.com/youtype/mypy_boto3_builder/releases)
- [Commits](https://github.com/youtype/mypy_boto3_builder/commits)

Updates `hypothesis` from 6.99.13 to 6.100.0
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-6.99.13...hypothesis-python-6.100.0)

Updates `moto` from 5.0.3 to 5.0.4
- [Release notes](https://github.com/getmoto/moto/releases)
- [Changelog](https://github.com/getmoto/moto/blob/master/CHANGELOG.md)
- [Commits](https://github.com/getmoto/moto/compare/5.0.3...5.0.4)

Updates `requests-mock` from 1.11.0 to 1.12.1
- [Release notes](https://github.com/jamielennox/requests-mock/releases)
- [Commits](https://github.com/jamielennox/requests-mock/compare/1.11.0...1.12.1)

Updates `types-jsonschema` from 4.21.0.20240311 to 4.21.0.20240331
- [Commits](https://github.com/python/typeshed/commits)

---
updated-dependencies:
- dependency-name: boto3-stubs
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: development-dependencies
- dependency-name: hypothesis
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: development-dependencies
- dependency-name: moto
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: development-dependencies
- dependency-name: requests-mock
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: development-dependencies
- dependency-name: types-jsonschema
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: development-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

WillDaSilva commented on src/meltano/core/project.py:
```diff
@@ -416,122 +423,157 @@ def dotenv_update(self):
 
     @makedirs
-    def meltano_dir(self, *joinpaths):
+    def meltano_dir(self, *joinpaths: StrPath, make_dirs: bool = True) -> Path:  # noqa: ARG002
```
The `make_dirs` parameter here and in some other functions appears unused. If it isn't supported but is included for the signature, or if it must be `True`, then a `NotImplementedError` or `ValueErrror` exception should be raised otherwise.

edgarrmondragon replied:
It's actually required so we can pass down the flag from methods like `Project.logs_dir`. For example, without it the following code fails

```python
from meltano.core.project import Project

project = Project.find(".")
project.logs_dir(make_dirs=False)
```

with

```python
TypeError: Project.meltano_dir() got an unexpected keyword argument 'make_dirs'
```

That's because all functions decorated with `@makedirs` are expected to accept the `make_dirs` keyword argument after the changes introduced to the decorator in this PR.

Wdyt?

WillDaSilva replied:
After reading through the code more closely, and understanding it better, my impression is that the `makedirs` decorator is a bad idea. Directory creation shouldn't be inherently tied to retrieving information about what those directories would be. That said, changing that goes beyond the scope of this PR, which provides a valuable fix.

dependabot[bot] committed:
chore(deps): bump the runtime-dependencies group with 7 updates (#8473)

* chore(deps): bump the runtime-dependencies group with 7 updates

Bumps the runtime-dependencies group with 7 updates:

| Package | From | To |
| --- | --- | --- |
| [boto3](https://github.com/boto/boto3) | `1.34.64` | `1.34.74` |
| [click-didyoumean](https://github.com/click-contrib/click-didyoumean) | `0.3.0` | `0.3.1` |
| [croniter](https://github.com/kiorky/croniter) | `2.0.2` | `2.0.3` |
| [google-cloud-storage](https://github.com/googleapis/python-storage) | `2.15.0` | `2.16.0` |
| [importlib-resources](https://github.com/python/importlib_resources) | `6.3.1` | `6.4.0` |
| [smart-open](https://github.com/piskvorky/smart_open) | `7.0.1` | `7.0.4` |
| [sqlalchemy](https://github.com/sqlalchemy/sqlalchemy) | `2.0.28` | `2.0.29` |


Updates `boto3` from 1.34.64 to 1.34.74
- [Release notes](https://github.com/boto/boto3/releases)
- [Changelog](https://github.com/boto/boto3/blob/develop/CHANGELOG.rst)
- [Commits](https://github.com/boto/boto3/compare/1.34.64...1.34.74)

Updates `click-didyoumean` from 0.3.0 to 0.3.1
- [Release notes](https://github.com/click-contrib/click-didyoumean/releases)
- [Commits](https://github.com/click-contrib/click-didyoumean/compare/v0.3.0...v0.3.1)

Updates `croniter` from 2.0.2 to 2.0.3
- [Changelog](https://github.com/kiorky/croniter/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/kiorky/croniter/commits)

Updates `google-cloud-storage` from 2.15.0 to 2.16.0
- [Release notes](https://github.com/googleapis/python-storage/releases)
- [Changelog](https://github.com/googleapis/python-storage/blob/main/CHANGELOG.md)
- [Commits](https://github.com/googleapis/python-storage/compare/v2.15.0...v2.16.0)

Updates `importlib-resources` from 6.3.1 to 6.4.0
- [Release notes](https://github.com/python/importlib_resources/releases)
- [Changelog](https://github.com/python/importlib_resources/blob/main/NEWS.rst)
- [Commits](https://github.com/python/importlib_resources/compare/v6.3.1...v6.4.0)

Updates `smart-open` from 7.0.1 to 7.0.4
- [Release notes](https://github.com/piskvorky/smart_open/releases)
- [Changelog](https://github.com/piskvorky/smart_open/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/piskvorky/smart_open/compare/v7.0.1...v7.0.4)

Updates `sqlalchemy` from 2.0.28 to 2.0.29
- [Release notes](https://github.com/sqlalchemy/sqlalchemy/releases)
- [Changelog](https://github.com/sqlalchemy/sqlalchemy/blob/main/CHANGES.rst)
- [Commits](https://github.com/sqlalchemy/sqlalchemy/commits)

---
updated-dependencies:
- dependency-name: boto3
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: runtime-dependencies
- dependency-name: click-didyoumean
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: runtime-dependencies
- dependency-name: croniter
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: runtime-dependencies
- dependency-name: google-cloud-storage
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: runtime-dependencies
- dependency-name: importlib-resources
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: runtime-dependencies
- dependency-name: smart-open
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: runtime-dependencies
- dependency-name: sqlalchemy
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: runtime-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>

* Fix tests

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Edgar Ramírez-Mondragón <edgar@meltano.com>

edgarrmondragon committed:
Merge branch 'main' into edgarrmondragon/fix/recursive-makedirs

edgarrmondragon committed:
Type check `meltano.core.project`
